Refactor `mergeFlags` into `config.MergeFlags`

Shikachuu opened the pull request:
- Move away from `cmd.mergeFlags` in favor of `config.MergeFlags` 
- Fix multiple equal sing handling for values in `config.MergeFlags` with tests
- Add start-end single quote removal to `config.MergeFlags` with tests
- Remove old `cmd.mergeFlags`
<!--- Provide a general summary of your changes in the Title above -->
Closes #592 
## Description
Handling of value single + double quotes and the `=` char was inconsistent and the quotes were included in the values string, this PR fixes that.

## Motivation and Context
`cmd.mergeFlags` was an exact copy of `config.MergeFlags` and multiple app used both.
It is generally a great idea to move commonly used functionality under `pkg` and remove duplicate code.
- [X] I have raised an issue to propose this change, which has been given a label of `design/approved` by a maintainer ([required](https://github.com/alexellis/arkade/blob/master/CONTRIBUTING.md)) #592 


## How Has This Been Tested?
All the changes covered by the extended unit test for the `config.MergeFlags` function.

Also, I ran every previously not working command from issue #592.
<details>
<summary>Logs</summary>
<pre><code>
Using Kubeconfig: /home/shika/.kube/config
Using Kubeconfig: /home/shika/.kube/config
[Warning] unable to create namespace default, may already exist: Error from server (AlreadyExists): namespaces "default" already exists
Client: x86_64, Linux
"ingress-nginx" already exists with the same configuration, skipping
VALUES values.yaml
Command: /home/shika/.arkade/bin/helm [upgrade --install ingress-nginx ingress-nginx/ingress-nginx --namespace default --values /tmp/charts/ingress-nginx/values.yaml --set controller.extraArgs.publish-status-address=localhost --set controller.updateStrategy.rollingUpdate.maxUnavailable=1 --set controller.config.log-format-upstream=remote_addr=$remote_addr user=$remote_user ts=$time_local request="$request" status=$status body_bytes=$body_bytes_sent referer="$http_referer" user_agent="$http_user_agent" request_length=$request_length duration=$request_time upstream=$proxy_upstream_name upstream_addr=$upstream_addr upstream_resp_length=$upstream_response_length upstream_duration=$upstream_response_time upstream_status=$upstream_status traceId=$opentracing_context_uber_trace_id --set controller.hostPort.enabled=true --set controller.service.nodePorts.http=30080 --set controller.publishService.enabled=false --set controller.updateStrategy.rollingUpdate.maxSurge=0 --set controller.config.enable-opentracing=true --set controller.config.jaeger-collector-host=tempo.default.svc.cluster.local --set controller.service.type=NodePort]
Release "ingress-nginx" does not exist. Installing it now.
NAME: ingress-nginx
LAST DEPLOYED: Wed Mar  2 22:54:20 2022
NAMESPACE: default
STATUS: deployed
REVISION: 1
TEST SUITE: None
NOTES:
The ingress-nginx controller has been installed.
Get the application URL by running these commands:
  export HTTP_NODE_PORT=30080
  export HTTPS_NODE_PORT=$(kubectl --namespace default get services -o jsonpath="{.spec.ports[1].nodePort}" ingress-nginx-controller)
  export NODE_IP=$(kubectl --namespace default get nodes -o jsonpath="{.items[0].status.addresses[1].address}")

  echo "Visit http://$NODE_IP:$HTTP_NODE_PORT to access your application via HTTP."
  echo "Visit https://$NODE_IP:$HTTPS_NODE_PORT to access your application via HTTPS."

An example Ingress that makes use of the controller:
  apiVersion: networking.k8s.io/v1
  kind: Ingress
  metadata:
    name: example
    namespace: foo
  spec:
    ingressClassName: nginx
    rules:
      - host: www.example.com
        http:
          paths:
            - backend:
                service:
                  name: exampleService
                  port:
                    number: 80
              path: /
    # This section is only required if TLS is to be enabled for the Ingress
    tls:
      - hosts:
        - www.example.com
        secretName: example-tls

If TLS is enabled for the Ingress, a Secret containing the certificate and key must also be provided:

  apiVersion: v1
  kind: Secret
  metadata:
    name: example-tls
    namespace: foo
  data:
    tls.crt: <base64 encoded cert>
    tls.key: <base64 encoded key>
  type: kubernetes.io/tls
=======================================================================
= ingress-nginx has been installed.                                   =
=======================================================================

# If you're using a local environment such as "minikube" or "KinD",
# then try the inlets operator with "arkade install inlets-operator"

# If you're using a managed Kubernetes service, then you'll find
# your LoadBalancer's IP under "EXTERNAL-IP" via:

kubectl get svc ingress-nginx-controller

# Find out more at:
# https://github.com/kubernetes/ingress-nginx/tree/master/charts/ingress-nginx

Thanks for using arkade!

</code></pre>
</details>

## Are you a GitHub Sponsor (Yes/No?)

<!--- Check at https://github.com/sponsors/alexellis         -->
<!--- Sponsors get priority because they support the project -->

- [ ] Yes
- [X] No

## Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [X] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Documentation

- [ ] I have updated the list of tools in README.md if (required) with `./arkade get -o markdown`
- [ ] I have updated the list of apps in README.md if (required) with `./arkade install --help`

## Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [X] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [X] I've read the [CONTRIBUTION](https://github.com/alexellis/arkade/blob/master/CONTRIBUTING.md) guide
- [X] I have signed-off my commits with `git commit -s`
- [X] I have added tests to cover my changes.
- [X] All new and existing tests passed.

<!--- In case it is a new application -->
- [X] I have tested this on arm, or have added code to prevent deployment

alexellis commented on cmd/apps/openfaas_app_test.go:
```diff
@@ -22,3 +23,77 @@ func Test_getValuesSuffix_aarch64(t *testing.T) {
+			},
+		},
+		{
+			"issue #592",
```
We could improve the title of this test case.

Shikachuu replied:
Sure, but I was thinking about remove test case naming all together. Should I? Or should I just rename the cases?

Shikachuu replied:
I kept the naming and this time they are meaningful ones!

alexellis commented on cmd/apps/openfaas_app_test.go:
```diff
@@ -22,3 +23,77 @@ func Test_getValuesSuffix_aarch64(t *testing.T) {
+		want  map[string]string
+	}{
+		{
+			"openfaas",
```
Likewise, this scenario is really just about the dash within a value, update the title and remove anything that isn't required like enabled=true etc.

alexellis commented on cmd/apps/openfaas_app_test.go:
```diff
@@ -22,3 +23,77 @@ func Test_getValuesSuffix_aarch64(t *testing.T) {
+				"nats.external.enabled=true",
+				"nats.external.host=openfaas-stan.openfaas.svc.cluster.local",
+				"nats.external.port=4222",
+				"nats.clusterName=\"faas-cluster\"",
```
Bash may not quote the string in the way you have here, I suspect it has no `\"` character at all.

Shikachuu replied:
In this case we are responsible for parsing the key/value paris, using plain SH it saddly appears in the go string.

Shikachuu replied:
You were right, all the double quotes removed in case of sh, bash and fish as well.
I removed the double quote removal from the function with the test cases.

alexellis commented on cmd/apps/openfaas_app_test.go:
```diff
@@ -22,3 +23,77 @@ func Test_getValuesSuffix_aarch64(t *testing.T) {
+		{
+			"issue #592",
+			[]string{
+				"controller.config.enable-opentracing='true'",
```
All this detail is not required to prove that the one thing that was broken is fixed.

Less is more, reduce as much as possible so we can read this and understand what we're maintaining if we ever have to revisit it.

alexellis commented on cmd/apps/openfaas_app.go:
```diff
@@ -229,10 +229,18 @@ func getValuesSuffix(arch string) string {
 		if len(flag) != 2 {
 			return fmt.Errorf("incorrect format for custom flag `%s`", setOverride)
 		}
+		isStartFlagsParsed := false
```
What does `isStartFlagsParsed` mean? I think there may be a better variable name that explains what this is doing.

Shikachuu replied:
Yeah at that point of time, a thought this is a-okay, now reading only this 3 rows I honestly have hard time to figure it out what this does. We've all been there :D I'll change this ASAP.

alexellis commented on cmd/apps/openfaas_app.go:
```diff
@@ -229,10 +229,18 @@ func getValuesSuffix(arch string) string {
 			return fmt.Errorf("incorrect format for custom flag `%s`", setOverride)
 		}
+		isStartFlagsParsed := false
+		if strings.HasPrefix(flag[1], "\"") && strings.HasSuffix(flag[1], "\"") {
```
I don't think bash will pass `\"` for you?

alexellis commented:
This could be related:

https://github.com/helm/helm/issues/1994

Shikachuu commented:
> > How Has This Been Tested?
> 
> We would need to see output from a successful installation, and objects with the data / strings populated inside K8s. Unit tests are not enough here, despite usually being useful or all that's required.

I ran the command that Lucas provided in the issue, attached the successfull install log to the base PR in a collapse.

Shikachuu commented:
@alexellis I changed the description and the title since I find some charts, that doesn't reflected the changes the changes from this PR like certmanager or Grafana.
Turned out there are 2 MergeFlags, one in the `pkg/config/config.go` and one in the `cmd/openfaas.go`.

I removed the one from `cmd/openfaas.go` and moved all the commands that used that to use the one from the `pkg` level, since that function has already had good quality, passing unit tests!
Also there was no reason to maintain 2 with the exact same functionality.
The new bug fixes were moved as well with the unit tests.

I know this PR is now chaged it's shape by a lot, but I hope it happened in the favor of quality and maintainability.
